Index: git/MediaSession.cpp
===================================================================
--- git.orig/MediaSession.cpp
+++ git/MediaSession.cpp
@@ -269,7 +269,7 @@ DRM_RESULT MediaKeySession::PolicyCallba
 {
     /*!+!hla fix this, implement for something. */
     DRM_RESULT dr = DRM_SUCCESS;
-    const DRM_PLAY_OPL_EX2 *oplPlay = NULL;
+    const DRM_PLAY_OPL_3_0 *oplPlay = NULL;
 
     BSTD_UNUSED(f_pKID);
     BSTD_UNUSED(f_pLID);
@@ -280,7 +280,7 @@ DRM_RESULT MediaKeySession::PolicyCallba
         case DRM_PLAY_OPL_CALLBACK:
             printf("  Got DRM_PLAY_OPL_CALLBACK from Bind:\r\n");
             ChkArg( f_pvPolicyCallbackData != NULL );
-            oplPlay = (const DRM_PLAY_OPL_EX2*)f_pvPolicyCallbackData;
+            oplPlay = (const DRM_PLAY_OPL_3_0*)f_pvPolicyCallbackData;
 
             printf("    minOPL:\r\n");
             printf("    wCompressedDigitalVideo   = %d\r\n", oplPlay->minOPL.wCompressedDigitalVideo);
Index: git/MediaSessionExt.cpp
===================================================================
--- git.orig/MediaSessionExt.cpp
+++ git/MediaSessionExt.cpp
@@ -84,17 +84,17 @@ DRM_RESULT opencdm_output_levels_callbac
     } 
 
     MediaKeySession::DecryptContext * const decryptContext = const_cast<MediaKeySession::DecryptContext *>(static_cast<const MediaKeySession::DecryptContext*>(data));
-    const DRM_PLAY_OPL_EX2 * const opl = static_cast<const DRM_PLAY_OPL_EX2 *>(outputLevels);
+    const DRM_PLAY_OPL_3_0 * const opl = static_cast<const DRM_PLAY_OPL_3_0 *>(outputLevels);
 
     ASSERT(opl->dwVersion == 0);
 
     decryptContext->outputProtection.setOutputLevels(opl->minOPL);
 
     // MaxRes Decode
-    const DRM_VIDEO_OUTPUT_PROTECTION_IDS_EX &dvopi = opl->dvopi;
+    const DRM_DIGITAL_VIDEO_OUTPUT_PROTECTION_IDS_3_0 &dvopi = opl->dvopi;
     for (size_t i = 0; i < dvopi.cEntries; ++i)
     {
-        const DRM_OUTPUT_PROTECTION_EX &dope = dvopi.rgVop[i];
+        const DRM_OUTPUT_PROTECTION_3_0 &dope = dvopi.rgVop[i];
         if (DRM_IDENTICAL_GUIDS(&dope.guidId, &g_guidMaxResDecode))
         {
             ASSERT(dope.dwVersion == 3);
@@ -225,7 +225,7 @@ CDMi_RESULT MediaKeySession::StoreLicens
     // response. There is always only a single BID per server response.
 
     // BID
-    mBatchId = drmLicenseResponse.m_oBatchID; 
+    mBatchId = drmLicenseResponse.m_idSession; 
     PrintBase64(sizeof(mBatchId.rgb), mBatchId.rgb, "BatchId/SecureStopId");
 
     // Microsoft says that a batch ID of all zeros indicates some sort of error
