[ARRISEOS-29522] Pause progressive Pipeline only when buffer is small

diff --git a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
index 65b816ff..75bfe69f 100644
--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -1947,7 +1947,10 @@ void MediaPlayerPrivateGStreamer::updateStates()
         } else if (state == GST_STATE_PLAYING) {
             m_paused = false;
 
-            if ((m_buffering && !isLiveStream()) || !m_playbackRate) {
+            WTF::Seconds minBufferedRange(2);
+            WTF::Seconds bufferedTime(durationMediaTime().toDouble() * m_bufferingPercentage / 100.0);
+            bool needPause = bufferedTime - WTF::Seconds(playbackPosition()) < minBufferedRange;
+            if ((m_buffering && !isLiveStream() && needPause) || !m_playbackRate) {
                 GST_DEBUG("[Buffering] Pausing stream for buffering.");
                 changePipelineState(GST_STATE_PAUSED);
             }
